[Submodule] Turning flash-attention integration into 3rd party submod

drisspg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144120


# Summary

### Sticky points

Cuda-graph rng handling has changed / deviated from original implementation. We will be left with a dangling 'offset' val and confusing naming due to BC

## Dependencies
- Flash PR: https://github.com/Dao-AILab/flash-attention/pull/1419


### Other Points
- The BC linter is complaining about losing generate.py and its functions which is not real BC surface
cc @albanD

Differential Revision: [D68502879](https://our.internmc.facebook.com/intern/diff/D68502879)

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.h:
```diff
@@ -29,7 +31,7 @@ mha_varlen_fwd(const at::Tensor &q,  // total_q x num_heads x head_size, total_q
                const at::Tensor &cu_seqlens_k,  // b+1
                std::optional<at::Tensor> &seqused_k, // b. If given, only this many elements of each batch element's keys are used.
-               std::optional<at::Tensor> &block_table_, // batch_size x max_num_blocks_per_seq
+               c10::optional<at::Tensor> &block_table_, // batch_size x max_num_blocks_per_seq
```
fix

Skylion007 commented on CMakeLists.txt:
```diff
@@ -876,6 +876,7 @@ cmake_dependent_option(
 # feature by default We dont currently document this feature because we don't
 # Suspect users building from source will need this
 add_definitions(-DFLASHATTENTION_DISABLE_ALIBI)
+add_definitions(-DFLASH_NAMESPACE=pytorch_flash)
```
```suggestion
add_definitions(-DFLASHATTENTION_DISABLE_SOFTCAP)
add_definitions(-DFLASH_NAMESPACE=pytorch_flash)
```
We do not support softcaps after all.

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update on "[Submodule] Turning flash-attention integration into 3rd party submod"



# Summary

### Sticky points

Cuda-graph rng handling has changed / deviated from original implementation. We will be left with a dangling 'offset' val and confusing naming due to BC

## Dependencies
- Flash PR: 


[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented on .gitmodules:
```diff
@@ -134,3 +134,6 @@
 	url = https://git.gitlab.arm.com/kleidi/kleidiai.git
+[submodule "third_party/flash-attention"]
+	path = third_party/flash-attention
+	url = https://github.com/drisspg/flash-attention.git
```
temp until I land the PR

drisspg committed:
Update

[ghstack-poisoned]

albanD reviewed (commented):
Sounds good!
Looking forward to being able to land this!

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

r-barnes commented on aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.h:
```diff
@@ -1,10 +1,11 @@
 #pragma once
 #include <cstddef>
 
+#include <namespace_config.h>
```
Do you need a separate config file here versus some other existing standard PyTorch config header?

drisspg replied:
So this config header is this one: https://github.com/Dao-AILab/flash-attention/pull/1419/files#diff-1271a158569ddafcee0391c481c061a54b79bd61dba1c32c0514a5182b4f4e75 which probably has more extras then needed

r-barnes commented on aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.h:
```diff
@@ -18,6 +19,7 @@ mha_fwd(const at::Tensor &q,         // batch_size x seqlen_q x num_heads x head
         bool is_causal,
         int window_size_left,
         int window_size_right,
+        const float softcap,
```
Why is this change happening at the same time? (Also, nice use of `const`!)

drisspg replied:
So this PR also brings us to a much newer version of flash that has since added these params, I could probably make a stack that adds this dummy arg if helpful for review

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update on "[Submodule] Turning flash-attention integration into 3rd party submod"



# Summary

### Sticky points

Cuda-graph rng handling has changed / deviated from original implementation. We will be left with a dangling 'offset' val and confusing naming due to BC

## Dependencies
- Flash PR: https://github.com/Dao-AILab/flash-attention/pull/1419


### Other Points
- The BC linter is complaining about losing generate.py and its functions which is not real BC surface
cc albanD

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@drisspg has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68502879).

drisspg commented on CMakeLists.txt:
```diff
@@ -876,6 +876,10 @@ cmake_dependent_option(
 # feature by default We dont currently document this feature because we don't
 # Suspect users building from source will need this
 add_definitions(-DFLASHATTENTION_DISABLE_ALIBI)
+add_definitions(-DFLASHATTENTION_DISABLE_SOFTCAP)
```
move these here:
https://github.com/pytorch/pytorch/blob/3ab6395c71fe7ea8f909fc318caf7cbe0e67e625/caffe2/CMakeLists.txt#L989
